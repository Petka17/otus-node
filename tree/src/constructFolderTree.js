import { promises as fs } from 'fs'
import path from 'path'

export const constructFolderTree = async (fullPath) => {
  const childrenList = await fs.readdir(fullPath)

  const items = await processChildren(childrenList, fullPath)

  return { name: `${path.basename(fullPath)}`, items }
}

const processChildren = async (childrenList, parentPath) =>
  Promise.all(
    childrenList.map(async (name) => {
      const folder = path.join(parentPath, name)
      const isDirectory = (await fs.lstat(folder)).isDirectory()

      if (!isDirectory) return { name }

      const childrenList = await fs.readdir(folder)
      const items = await processChildren(childrenList, folder)

      return { name: `${name}/`, items }
    }),
  )
