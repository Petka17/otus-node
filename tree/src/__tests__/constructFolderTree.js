import { constructFolderTree } from '../constructFolderTree.js'
import testData from '../__test-data__/testDirTree.js'

jest.mock('path', () => {
  const testData = require('../__test-data__/testDirTree.js')

  return {
    basename: jest.fn(() => testData.name),
    join: jest.fn((...args) => args.join('/')),
  }
})

const mockGetNode = (path, items) => {
  const head = path[0]
  const tail = path.slice(1)

  const item = items.find((item) => item.name === head)

  if (!item) return { name: 'not_correct_path', items: [] }

  if (tail.length === 0) return item

  return mockGetNode(tail, item.items)
}

jest.mock('fs', () => {
  const testData = require('../__test-data__/testDirTree.js')

  const getItems = (folder) =>
    mockGetNode(folder.slice(testData.fullPath.length + 1).split('/'), [testData]).items

  return {
    promises: {
      readdir: jest.fn((folder) => {
        const items = getItems(folder)
        return Promise.resolve(Array.isArray(items) ? items.map((item) => item.name) : [])
      }),
      lstat: jest.fn((folder) => {
        return Promise.resolve({
          isDirectory: () => (getItems(folder) ? true : false),
        })
      }),
    },
  }
})

test('Construct Folder Tree works', async (done) => {
  const strTree = await constructFolderTree(`${testData.fullPath}/${testData.name}`)

  expect(strTree).toMatchSnapshot()

  done()
})
