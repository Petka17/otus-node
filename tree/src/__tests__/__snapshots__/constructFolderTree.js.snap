// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Construct Folder Tree works 1`] = `
Object {
  "items": Array [
    Object {
      "items": Array [
        Object {
          "name": "child11",
        },
      ],
      "name": "child1/",
    },
    Object {
      "name": "child2",
    },
    Object {
      "items": Array [],
      "name": "child3/",
    },
  ],
  "name": "root",
}
`;
